File: 2021/csharp/Aoc2021/Day02.cs
namespace Aoc2021;

public class Day02 : IDay
{
    public override string Day { get; } = nameof(Day02)[3..];

    public override string Result1()
    {
        List<Command> commands = ParseCommands(InputLines);
        Submarine submarine = MoveFirstTry(new(0, 0, 0), commands);
        int result = submarine.HorizontalPosition * submarine.Depth;
        return result.ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        List<Command> commands = ParseCommands(InputLines);
        Submarine submarine = MoveSecondTry(new(0, 0, 0), commands);
        int result = submarine.HorizontalPosition * submarine.Depth;
        return result.ToString(CultureInfo.InvariantCulture);
    }

    public record Submarine(int HorizontalPosition, int Depth, int Aim);
    public record Command(string Direction, int Value);

    public static List<Command> ParseCommands(IEnumerable<string> input)
    {
        return input.Select((line) =>
        {
            string[] chunks = line.Split(' ', 2);
            return new Command(chunks[0], ParseInt(chunks[1]));
        }).ToList();
    }

    public static Submarine MoveFirstTry(Submarine submarine, IEnumerable<Command> commands)
    {
        return commands
            .Aggregate(submarine, (submarine, command) =>
            {
                (int horizontalPosition, int depth, _) = submarine;
                (string direction, int value) = command;
                return direction switch
                {
                    "forward" => submarine with { HorizontalPosition = horizontalPosition + value },
                    "down" => submarine with { Depth = depth + value },
                    "up" => submarine with { Depth = depth - value },
                    _ => throw new ArgumentException("Invalid command direction"),
                };
            });
    }

    public static Submarine MoveSecondTry(Submarine submarine, IEnumerable<Command> commands)
    {
        return commands
            .Aggregate(submarine, (submarine, command) =>
            {
                (int horizontalPosition, int depth, int aim) = submarine;
                (string direction, int value) = command;
                return direction switch
                {
                    "forward" => submarine with
                    {
                        HorizontalPosition = horizontalPosition + value,
                        Depth = depth + (aim * value),
                    },
                    "down" => submarine with { Aim = aim + value },
                    "up" => submarine with { Aim = aim - value },
                    _ => throw new ArgumentException("Invalid command direction"),
                };
            });
    }
}

File: 2021/fsharp/Aoc2021/Day02.fs
module Day02

open Utils

type Submarine = {
    HorizontalPosition: int
    Depth: int
    Aim: int
}

type Command = { Direction: string; Value: int }

let parseCommands (lines: string seq) =
    lines
    |> Seq.map (fun line ->
        let chunks = line.Split(' ', 2)

        {
            Direction = chunks.[0]
            Value = (chunks.[1] |> int)
        })
    |> Seq.toList

let moveFirstTry submarine commands =
    commands
    |> List.fold
        (fun submarine command ->
            let {
                    HorizontalPosition = horizontalPosition
                    Depth = depth
                } =
                submarine

            let { Direction = direction; Value = value } = command

            match direction with
            | "forward" -> {
                submarine with
                    HorizontalPosition = horizontalPosition + value
              }
            | "down" -> { submarine with Depth = depth + value }
            | "up" -> { submarine with Depth = depth - value }
            | _ -> failwith "Invalid command direction")
        submarine

let moveSecondTry submarine commands =
    commands
    |> List.fold
        (fun submarine command ->
            let {
                    HorizontalPosition = horizontalPosition
                    Depth = depth
                    Aim = aim
                } =
                submarine

            let { Direction = direction; Value = value } = command

            match direction with
            | "forward" -> {
                submarine with
                    HorizontalPosition = horizontalPosition + value
                    Depth = depth + (aim * value)
              }
            | "down" -> { submarine with Aim = aim + value }
            | "up" -> { submarine with Aim = aim - value }
            | _ -> failwith "Invalid command direction")
        submarine


let input = readInputLines "02" |> parseCommands

let result (submarine: Submarine) =
    submarine.HorizontalPosition * submarine.Depth

let job1 () =
    input
    |> moveFirstTry {
        HorizontalPosition = 0
        Depth = 0
        Aim = 0
    }
    |> result
    |> string

let job2 () =
    input
    |> moveSecondTry {
        HorizontalPosition = 0
        Depth = 0
        Aim = 0
    }
    |> result
    |> string
